src/csharp/Microsoft.Spark/Utils/UdfSerDe.cs

```diff
@@ -184,30 +344,62 @@ private static TypeData SerializeType(Type type)
 
         private static Type DeserializeType(TypeData typeData) =>
             s_typeCache.GetOrAdd(typeData,
-                td => LoadAssembly(typeData.ManifestModuleName).GetType(typeData.Name));
+                td => LoadAssembly(td.AssemblyName, td.ManifestModuleName).GetType(td.Name));
+
+        /// <summary>
+        /// Return the cached assembly, otherwise attempt to load and cache the assembly
+        /// in the following order:
+        /// 1) Search the assemblies loaded in the current app domain.
+        /// 2) Load the assembly from disk using manifestModuleName.
+        /// </summary>
+        /// <param name="assemblyName">The full name of the assembly</param>
+        /// <param name="manifestModuleName">Name of the module that contains the assembly</param>
+        /// <returns>Cached or Loaded Assembly</returns>
+        private static Assembly LoadAssembly(string assemblyName, string manifestModuleName)
+        {
+            return s_assemblyCache.GetOrAdd(
+                assemblyName,
+                _ =>
+                {
+                    foreach (Assembly asm in AppDomain.CurrentDomain.GetAssemblies())
+                    {
+                        if (asm.FullName.Equals(assemblyName))
+                        {
+                            return asm;
+                        }
+                    }
+
+                    return LoadAssembly(manifestModuleName);
```

Comment:
can we add a comment this this can potentially load multiple dlls with the same name? (or we can fix the race condition in this PR)

Reply:
fixed race condition using Lazy<Assembly>().Value

Reply:
great!